### ISSUE TITLE
UART receiver and transmitter IP

### ISSUE THREAD

**Issue by eiffel-fl** [echopen/PRJ-medtech_embsys#2]
The goal is to develop two [IPs](https://en.m.wikipedia.org/wiki/Semiconductor_intellectual_property_core) for [UART](https://en.m.wikipedia.org/wiki/Universal_asynchronous_receiver/transmitter) :
* one for reception,
* the other for transmission

Indicative deadline : 15~30 may

**Comment by sprotch** [comment]
What is the goal here? To implement UART controllers in VHDL for the programmable logic part of the RedPitaya? Why do we not use the HW controllers in the Zynq PS?

**Comment by eiffel-fl** [comment]
The goal is to implement UART in Verilog to permit this [card](https://www.digikey.fr/product-detail/fr/lattice-semiconductor-corporation/ICE40HX8K-B-EVN/220-1874-ND/4738851) to send and receive data from the PC it is connected.

**Comment by sprotch** [comment]
A full reference design for a 16550 UART on Lattice ICE40 already exists: http://www.latticesemi.com/en/Products/DesignSoftwareAndIP/IntellectualProperty/ReferenceDesigns/ReferenceDesigns01/UART16550Transceiver.aspx
If you don't want to go with this design, have a look at OpenCores (opencores.org). Several open source cores already exist.
And if none of the above fits your needs, I would advise to go with a single core/IP and not two separate ones (one for TX and the other for TX) to make synchronization easier and implement a real UART in a single module.

**Comment by eiffel-fl** [comment]
I know about opencores.org and I check some design but I want to design mine.

I will try to make only one IP if you think that it is better.

**Comment by eiffel-fl** [comment]
I am encountering problem with the tool to synthesize the TX IP.
When I test the code which is synthesized in my testbench it seems that the state registers are 'X' which seems impossible.

I will investigate this but as I am in holiday next week I will not be able to finish for the 30th May, so the new indicative date will be 15th-30th June.

**Comment by sprotch** [comment]
Do you have your code committed somewhere so that others can review and perhaps help to find the issue?

**Comment by eiffel-fl** [comment]
Yes [here](https://gitlab.com/eiffel/PRJ-medtech_embsys/tree/master/IPs/UART/TX/verilog) .

I reviewed it and I did not find something aberrant.
I think I will change the design of the TX IP hoping it will help.
